internal/states/statemgr/filesystem.go

```diff
@@ -92,6 +92,16 @@ func NewFilesystemBetweenPaths(readPath, writePath string) *Filesystem {
 	}
 }
 
+// SaveErroredTestStateFile saves the state present in memory as a stateFile in disk
+// and is invoked when "tofu test" fails to destroy resources.
+func SaveErroredTestStateFile(stateFilePath string, state *states.State) error {
+	log.Printf("Writing state to file: errored_test.tfstate")
+	fs := NewFilesystem(stateFilePath)
+	schema := &tofu.Schemas{}
+	err := WriteAndPersist(fs, state, schema)
```

Comment:
I'm concerned here that WriteState ends up being called here and this will attempt to refresh the state. https://github.com/opentffoundation/opentf/blob/b66111f8bf80d0bdb92684f9287e0751487902f6/internal/states/statemgr/filesystem.go#L139-L144

I have a feeling we should be calling `WriteStateForMigration` similar to how the errored state writing works.

Is there any chance we could re-work this PR to use code similar to [this](https://github.com/opentffoundation/opentf/blob/cb2e9119aa75eeb8e1fa175e2b7a205a4fef129f/internal/backend/local/backend_apply.go#L284-L335) (or extract this code out to a re-usable method where you can pass different filenames?

Reply:
> I have a feeling we should be calling `WriteStateForMigration` similar to how the errored state writing works.
> 
> Is there any chance we could re-work this PR to use code similar to [this](https://github.com/opentffoundation/opentf/blob/cb2e9119aa75eeb8e1fa175e2b7a205a4fef129f/internal/backend/local/backend_apply.go#L284-L335) (or extract this code out to a re-usable method where you can pass different filenames?

I am a little unsure of how do you want me to change this part. Do you want me to write a function similar to `backupStateForError`, where it attempts to write a backup for `errored.tfstate`, and even that fails it logs out the state file to the console as a last ditch attempt. Or should I only make a call to `WriteStateForMigration` where the write operation occurs without an attempt to refresh. Please do let me know,  I have made the other changes requested and will push changes after this. Thanks!

Reply:
You're right, and I missed that WriteStateForMigration calls refresh too. Maybe this isn't so bad, but I've not tracked down yet what is passed around without debugging this.

In my original comment I asked

> Is there any chance we could re-work this PR to use code similar to [this](https://github.com/opentffoundation/opentf/blob/cb2e9119aa75eeb8e1fa175e2b7a205a4fef129f/internal/backend/local/backend_apply.go#L284-L335) (or extract this code out to a re-usable method where you can pass different filenames?

And I think this still stands. Mainly for the following reasons:

1. The code linked has great diagnostic logging and error handling
2. The code linked just has a simple call out to `local.WriteStateForMigration` which your code can also use.
3. I don't see the point in introducing a new method that has a cut down version of the logic when the code already exists in some form (that we could either reuse, abstract out or base new work on top of)

Reply:
> Is there any chance we could re-work this PR to use code similar to [this](https://github.com/opentffoundation/opentf/blob/cb2e9119aa75eeb8e1fa175e2b7a205a4fef129f/internal/backend/local/backend_apply.go#L284-L335) (or extract this code out to a re-usable method where you can pass different filenames?

Hi, here is what I have tried:
- I went ahead and tried extracting the code to a method, to which I would pass a filename and it would handle storing the state to the file and associated diagnostics and error handling. 

- I was unsure of where to place this method, I initially placed it in `filesystem.go` [https://github.com/opentofu/opentofu/blob/cb2e9119aa75eeb8e1fa175e2b7a205a4fef129f/internal/states/statemgr/filesystem.go](url), however that gave me a cyclic dependency with package imports. Finding a suitable place to write this method would solve this problem, I would like your insights on this.

- I am going ahead and replicating the contents of 
`backupStateForError` https://github.com/opentofu/opentofu/blob/cb2e9119aa75eeb8e1fa175e2b7a205a4fef129f/internal/backend/local/backend_apply.go#L290-L334
in `DestorySummary` https://github.com/opentofu/opentofu/blob/cb2e9119aa75eeb8e1fa175e2b7a205a4fef129f/internal/command/views/test.go#L233

I personally don't like the idea of having similar code in two places, I would much rather have a method that can accept the filename as a parameter. Please do let me know of your opinions on how to tackle this.